### PR TITLE
Limit the received content when handling the content as a String.

### LINKED ISSUE #27293: Denial of service via excessive response size from identity provider
### Source: https://issues.redhat.com/browse/RHSSO-2890

Requests to identity providers have no response size limitation. Receiving a large HTTP response body leads to resource exhaustion and Keycloak crashes. For an exploit, a request must be triggered against a HTTP server under the control of an attacker, which sends a large HTTP response body.

**Requirements to exploit:**
Administrator rights, or a foreign IDP is controlled by another party

**Component affected:** 
org.keycloak.broker

**Version affected:** 
>=23.0.4

**Impact:** Low (Weakness). Considering that it requires administrative
access, or a malicious IDP controlled by another party, it should be
possible to mitigate the issue by blocking anomalously large responses, or
introducing a rate limiting in front of the Keycloak server.

**Steps to reproduce:**

1. Create and start a TCP server, which sends an infinite HTTP chunked
stream (see attached reproducer).

2. Setup two realms, one as a identity provider for the other ("Keycloak OpenID Connect"). Both client and IDP contain valid configuration, except "User Info URL" of the IDP configuration, which is set to the Python server URL. The realm, which acts as IDP should contain an example user.

3. Login as example user via IDP.

Keycloak crashes with:
keycloak-1 | Terminating due to java.lang.OutOfMemoryError: Java heap space
keycloak-1 exited with code 3

### PR DESCRIPTION
**Pull request by ahus1**
Closes #27293

### CODE REVIEW

**Comment by ahus1** [comment]
@rmartinc - as I've been working on the other issue, I had an idea on how to limit the response size without moving to v5 of the library. 

The maximum response size can be configured for `DefaultHttpClientFactory`, but all users of `SimpleHttp` would set their own maximum size. I don't have a good idea where to pull this information from. 

Maybe it is good enough for now? WDYT?

**Comment by ahus1** [comment]
@rmartinc - I've updated this PR. Please let me know if you think this is a valid approach when you have the time. 

Pinging also @abstractj as you've been managing the original issue.

File: server-spi-private/src/main/java/org/keycloak/connections/httpclient/SafeInputStream.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public int read(byte[] b, int off, int len) throws IOException {
+        int sizeRead = delegate.read(b, off, len);
+        bytesConsumed += sizeRead;
+        checkConsumedBytes();
+        return sizeRead;
+    }
+
+    private void checkConsumedBytes() throws IOException {
+        if (bytesConsumed >= maxBytesToConsume) {
+            throw new IOException(String.format("Response is longer than %s, with max bytes to be consumed being %d", bytesConsumed, maxBytesToConsume));
```
Comment (rmartinc) [inline server-spi-private/src/main/java/org/keycloak/connections/httpclient/SafeInputStream.java:49]: Just a minor thing, but this wouldn't be `>` insteadof `>=`, just to allow exact number. And I would use: ` Response is at least %s bytes in size, with max...`.

File: server-spi-private/src/main/java/org/keycloak/connections/httpclient/SafeInputStream.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    public SafeInputStream(InputStream delegate, long maxBytesToConsume) {
+        this.delegate = delegate;
+        this.maxBytesToConsume = maxBytesToConsume;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int sizeRead = delegate.read(b, off, len);
+        bytesConsumed += sizeRead;
```
Comment (rmartinc) [inline server-spi-private/src/main/java/org/keycloak/connections/httpclient/SafeInputStream.java:42]: Better do the same than in single read: `if (sizeRead > 0) bytesConsumed += sizeRead;`

File: server-spi-private/src/test/java/org/keycloak/broker/provider/util/SimpleHttpTest.java [hunk 3]
```diff
@@ -64,7 +65,7 @@ public static Collection<Object[]> entities() {
         @Test
         public void withCharset() throws IOException {
             HttpResponse httpResponse = createBasicResponse(entity);
-            SimpleHttp.Response response = new SimpleHttp.Response(httpResponse);
+            SimpleHttp.Response response = new SimpleHttp.Response(httpResponse, HttpClientProvider.DEFAULT_MAX_CONSUMED_RESPONSE_SIZE);
```
Comment (rmartinc) [inline server-spi-private/src/test/java/org/keycloak/broker/provider/util/SimpleHttpTest.java:68]: I think we could add a test here too. I did something quick in this commit: https://github.com/rmartinc/keycloak/commit/1e6562339d8296c77f6a84fba2181be9f198c2ff
You can use it or something similar.

**Review by rmartinc — CHANGES_REQUESTED** [review]
Very good job @ahus1!
Just some minor things that maybe you can consider to add.

**Review by rmartinc — APPROVED** [review]
Thanks for the changes! Good idea limiting `SimpleHttp` to the methods that pass the session. LGTM now! Let's see the tests!

**Review by abstractj — APPROVED** [review]

**Review by abstractj — DISMISSED** [review]

**Comment by ahus1** [comment]
@abstractj - I needed to rebase the PR due to conflicts in the documentation. Please re-approve when you have the time. Thanks!

**Review by rmartinc — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Pull request merged**